crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,499 @@
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`__all__` is not alphabetically sorted")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Sort `__all__` alphabetically".to_string())
+    }
+}
+
+pub(crate) fn sort_dunder_all_assign(
+    checker: &mut Checker,
+    node: &ast::StmtAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAssign { value, targets, .. } = node;
+    let [ast::Expr::Name(ast::ExprName { id, .. })] = targets.as_slice() else {
+        return;
+    };
+    sort_dunder_all(checker, id, value, parent);
+}
+
+pub(crate) fn sort_dunder_all_augassign(
+    checker: &mut Checker,
+    node: &ast::StmtAugAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAugAssign {
+        value,
+        target,
+        op: ast::Operator::Add,
+        ..
+    } = node
+    else {
+        return;
+    };
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+        return;
+    };
+    sort_dunder_all(checker, id, value, parent);
+}
+
+pub(crate) fn sort_dunder_all_annassign(
+    checker: &mut Checker,
+    node: &ast::StmtAnnAssign,
+    parent: &ast::Stmt,
+) {
+    let ast::StmtAnnAssign {
+        target,
+        value: Some(val),
+        ..
+    } = node
+    else {
+        return;
+    };
+    let ast::Expr::Name(ast::ExprName { id, .. }) = target.as_ref() else {
+        return;
+    };
+    sort_dunder_all(checker, id, val, parent);
+}
+
+fn sort_dunder_all(checker: &mut Checker, target: &str, node: &ast::Expr, parent: &ast::Stmt) {
+    if target != "__all__" {
+        return;
+    }
+
+    // We're only interested in `__all__` in the global scope
+    if !checker.semantic().current_scope().kind.is_module() {
+        return;
+    }
+
+    let locator = checker.locator();
+
+    let Some(dunder_all_val) = DunderAllValue::from_expr(node, locator) else {
+        return;
+    };
+
+    let new_dunder_all =
+        match dunder_all_val.construct_sorted_all(locator, parent, checker.stylist()) {
+            SortedDunderAll::AlreadySorted => return,
+            SortedDunderAll::Sorted(value) => value,
+        };
+
+    let dunder_all_range = dunder_all_val.range();
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, dunder_all_range);
+
+    if let Some(new_dunder_all) = new_dunder_all {
+        let applicability = {
+            if dunder_all_val.multiline
+                && checker.indexer().comment_ranges().intersects(node.range())
+            {
+                Applicability::Unsafe
+            } else {
+                Applicability::Safe
+            }
+        };
+        diagnostic.set_fix(Fix::applicable_edit(
+            Edit::range_replacement(new_dunder_all, dunder_all_range),
+            applicability,
+        ));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+struct DunderAllValue<'a> {
+    items: Vec<DunderAllItem>,
+    range: &'a TextRange,
```

Comment:
This is a good instinct, but `TextRange` is small and implements `Copy`, so it's preferable to just use `TextRange` here. (You won't even need to `.clone()` it.)

Reply:
I keep on thinking today will be the day I get to use lifetimes properly for the first time 🙃